File: src/main/java/dm/sql/executor/Formatter.java
package dm.sql.executor;

import java.io.PrintStream;
import java.sql.ResultSet;
import java.sql.SQLException;

public abstract class Formatter {

    protected final PrintStream printStream;

    public Formatter(PrintStream printStream) {
        this.printStream = printStream;
    }

    public abstract void out(ResultSet resultSet) throws SQLException;

    public abstract void out(int count) throws SQLException;

    public static class DsvFormatter extends Formatter {

        private final String rowSeperator;
        private final String columnSeperator;

        public DsvFormatter(PrintStream printStream, String rowSeperator, String columnSeparator) {
            super(printStream);
            this.rowSeperator = rowSeperator;
            this.columnSeperator = columnSeparator;
        }

        @Override
        public void out(ResultSet resultSet) throws SQLException {
            int i = resultSet.getMetaData().getColumnCount();
            while (resultSet.next()) {
                String[] values = new String[i];
                for (int i1 = 0; i1 < i; i1++) {
                    values[i1] = resultSet.getString(i1 + 1);
                }
                printStream.print(String.join(rowSeperator, values));
                printStream.print(columnSeperator);
            }
        }

        @Override
        public void out(int count) throws SQLException {
            printStream.print(count);
        }
    }

    public static class CsvFormatter extends DsvFormatter {

        public CsvFormatter(PrintStream printStream) {
            super(printStream, ",", "\n");
        }
    }

    public static class TsvFormatter extends DsvFormatter {

        public TsvFormatter(PrintStream printStream) {
            super(printStream, "\t", "\n");
        }
    }
}

File: src/main/kotlin/dm/sql/executor/Formatter.kt
package dm.sql.executor

import java.io.PrintStream
import java.sql.ResultSet

abstract class Formatter(var printStream: PrintStream) {

    abstract fun out(resultSet: ResultSet)

    abstract fun out(count: Int)

    open class DsvFormatter(printStream: PrintStream, var rowSeperator: String, private var columnSeparator: String) : Formatter(printStream) {

        override fun out(resultSet: ResultSet) {
            val i = resultSet.metaData.columnCount
            while (resultSet.next()) {

                val values = Array(i, {""})


                for (i1 in 0..i) {
                    values[i1] = resultSet.getString(i1 + 1)
                }

                printStream.print(values.joinToString(rowSeperator))
                printStream.print(columnSeparator)
            }
        }

        override fun out(count: Int) {
            printStream.print(count)
        }
    }

    class CsvFormatter(printStream: PrintStream) : DsvFormatter(printStream, ",", "\n")

    class TsvFormatter(printStream: PrintStream) : DsvFormatter(printStream, "\t", "\n")
}
